libbpf-tools/trace_helpers.c

```diff
@@ -990,6 +991,28 @@ bool is_kernel_module(const char *name)
 	return found;
 }
 
+static bool fentry_can_attach(int id)
+{
+	struct bpf_insn insns[] = { { .code = BPF_JMP | BPF_EXIT } };
+	struct bpf_prog_load_opts opts;
+	int prog_fd, attach_fd;
+
+	opts.sz = sizeof(opts);
+	opts.expected_attach_type = BPF_TRACE_FENTRY;
+	opts.attach_btf_id = id;
+
+	prog_fd = bpf_prog_load(BPF_PROG_TYPE_TRACING, "test", NULL, insns, 1, &opts);
+	if (prog_fd < 0)
+		return false;
+
+	attach_fd = bpf_raw_tracepoint_open(NULL, prog_fd);
+	if (attach_fd >= 0)
+		close(attach_fd);
+
+	close(prog_fd);
+	return attach_fd >= 0;
+}
+
 bool fentry_exists(const char *name, const char *mod)
```

Comment:
Since this change tries to load and attach a program with `BPF_TRACE_FENTRY` attach type, can we remove the searching done elsewhere in this function for `bpf_attach_type` enum and its `BPF_TRACE_FENTRY` member?

Reply:
Also, since the helper function you're adding is doing more significant / invasive checking, perhaps renaming `fentry_exists` to `fentry_can_attach` would make what's being done more obvious to callers, especially if the "test for existence of enum member" logic can be removed.

Could then call the function you're adding something like `fentry_try_attach` or merge the two functions - I don't feel strongly about this though.

Reply:
Done - removed the search and renamed to `fentry_can_attach` and `fentry_try_attach`.